### PR TITLE
[DRAFT] Implementation of the class feature

### PR DESCRIPTION
**Pull request by tfukaza**
Implementation of the class feature on the go back-end. The overall design closely resembles the Js version. More specifically:

- Class struct (in `class.go`)
- Handlers for requests related to classes (`classManagement.go`)
- Additional functions in db.go to communicate class-related information with Firebase
- Generation of a 3-word hash (tools/tinycrypt) for a given pid/cid, which will be stored in Firebase.
 
Note that the other PR that is currently active (#6) is the same as this PR, except it has the distributed counters (for hash generation) implemented as well. In this version, the 3-word hash is generated simply by using Go's built-in random number generator.  Be warned, as this may mean that we need to update `dist-counter`'s 2-word hash in the future to be backwards-compatible with the current version's 3-word hash.   

Lastly, though this has been tested (as seen in `server_test.go`) ideally it should get additional testing to ensure everything works, especially since the `tinycrypt` library was updated recently to work better with distributed counters. 

**IMPORTANT**
As stated in `constants.go`, you will need to make a new collection in Firebase called 'classes_alias' in order for this to work.

### CODE REVIEW

File: db/classManagement.go [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+		http.Error(w, "Error occurred in reading UID.", http.StatusInternalServerError)
+		return
+	}
+	
+	if req.Name == "" {
+		http.Error(w, "Error occurred in reading Name.", http.StatusInternalServerError)
+		return
+	}
+
+	if req.Thumbnail < 0 || req.Thumbnail >= 50 {
```
Comment (krashanoff) [inline db/classManagement.go:40]: We have a thumbnail count constant defined already in `constants.go` if I remember correctly. Make sure to use that one, unless there is an alternative thumbnail limit in the JavaScript backend.

File: db/classManagement.go [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+	c, err := d.GetClass(r.Context(), cid)
+
+	// check for error
+	if err != nil || c == nil {
+		http.Error(w, "Failed to get class (class does not exist or failed to unmarshal data)", http.StatusNotFound)
+		return
+	}
+
+	//check if the uid exists in the members list or instructor list
+	var is_in bool = false;
```
Comment (krashanoff) [inline db/classManagement.go:133]: I noticed that you are slipping into a more JavaScript-y style here. Remember that you can always declare it as `is_in := false`

File: db/db.go [hunk 3]
```diff
@@ -160,3 +163,169 @@ func (d *DB) DeleteProgram(ctx context.Context, uid string) error {
...
+	if err := doc.DataTo(c); err != nil {
+		return nil, err
+	}
+	
+	return c, err
+}
+
+// MakeAlias takes an id (usually pid or cid), generates a 3 word id(wid), and 
+// stores it in Firebase. The generated wid is returned as a string, with words comma seperated
+func (d *DB) MakeAlias(ctx context.Context, uid string, path string) (string, error) {
```
Comment (krashanoff) [inline db/db.go:274]: Nice work on MakeAlias! The word hash feature will be very nice to have.

**Comment by malsf21** [comment]
Looks like you're derefencing null somehow from the CI - not entirely sure what's happening but definitely investigate before merging! 

Otherwise, excited to see how this turns out 😄

**Comment by tfukaza** [comment]
> Looks like you're dereferencing null somehow from the CI - not entirely sure what's happening but definitely investigate before merging!
> 
> Otherwise, excited to see how this turns out 😄

I'm guessing it has to do with the firebase connection, as I get the same error when I test it locally, but when I supply my secret key it gets resolved.

**Comment by malsf21** [comment]
If it's just a key error, we can probably include that as a secret into the GitHub Actions tester - ideally a separate key just used for testing, with reduced perms.

Alternatively, there should maybe be a graceful failure point if the key doesn't exist? But that's probably out of the scope of this PR. I'd defer to @krashanoff too, there might be a better solution I haven't thought of yet!

**Comment by tfukaza** [comment]
> If it's just a key error, we can probably include that as a secret into the GitHub Actions tester - ideally a separate key just used for testing, with reduced perms.
> 
> Alternatively, there should maybe be a graceful failure point if the key doesn't exist? But that's probably out of the scope of this PR. I'd defer to @krashanoff too, there might be a better solution I haven't thought of yet!

Ye if you look at the README for this repo, it says not having a key will result in an error.

**Comment by malsf21** [comment]
> Ye if you look at the README for this repo, it says not having a key will result in an error.

It's a runtime goroutine error though, which is why I'm a bit confused 😮 . You can checkout the error log at [https://github.com/uclaacm/teach-la-go-backend/pull/9/checks?check_run_id=560844036](the actions link)!

**Comment by krashanoff** [comment]
It is because the tester lacks credentials to firebase. This is an issue I planned to address following the update to 1.14 pending this PR.

**Pull request merged**